### PR TITLE
restrict cell patches to locally owned cells

### PR DESCRIPTION
**Pull request by guidokanschat**

### CODE REVIEW

File: include/deal.II/dofs/dof_tools.h [hunk 1]
```diff
@@ -1699,18 +1699,24 @@ namespace DoFTools
...
-   * Create an incidence matrix that for every cell on a given level of a
-   * multilevel DoFHandler flags which degrees of freedom are associated with
-   * the corresponding cell. This data structure is a matrix with as many rows
-   * as there are cells on a given level, as many columns as there are degrees
-   * of freedom on this level, and entries that are either true or false. This
-   * data structure is conveniently represented by a SparsityPattern object.
-   *
-   * @note The ordering of rows (cells) follows the ordering of the standard
-   * cell iterators.
+   * Creates a block list in form of a sparsity pattern, which lists
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: Can you explain the term "block list"?

File: include/deal.II/dofs/dof_tools.h [hunk 2]
```diff
@@ -1699,18 +1699,24 @@ namespace DoFTools
...
+   *
+   * <tt>selected_dofs</tt> is a vector indexed by the local degrees
+   * of freedom on a cell. If it is used, only such dofs are entered
+   * into the block, which are selected. This allows for instance the
+   * exclusion of components or of dofs on the boundary.
    */
-  template <typename DoFHandlerType, class SparsityPatternType>
-  void make_cell_patches(SparsityPatternType     &block_list,
+  template <typename DoFHandlerType>
+  void make_cell_patches(SparsityPattern         &block_list,
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: It seems like a regression to only allow `SparsityPattern` here. Any reason?

File: include/deal.II/dofs/dof_tools.h [hunk 3]
```diff
@@ -1699,18 +1699,24 @@ namespace DoFTools
    */
   //@{
   /**
-   * Create an incidence matrix that for every cell on a given level of a
-   * multilevel DoFHandler flags which degrees of freedom are associated with
-   * the corresponding cell. This data structure is a matrix with as many rows
-   * as there are cells on a given level, as many columns as there are degrees
-   * of freedom on this level, and entries that are either true or false. This
-   * data structure is conveniently represented by a SparsityPattern object.
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: I thought this text is actually more descriptive of what I think the function does. Why delete it?

File: include/deal.II/dofs/dof_tools.h [hunk 4]
```diff
@@ -1699,18 +1699,24 @@ namespace DoFTools
...
-   *
-   * @note The ordering of rows (cells) follows the ordering of the standard
-   * cell iterators.
+   * Creates a block list in form of a sparsity pattern, which lists
+   * the degrees of freedom associated to each cell on the given
+   * level. The row index in this pattern is the cell index resulting
+   * from standard iteration through the Triangulation. For a
+   * parallel::distributed::Triangulation, only locally owned cells
+   * are entered.
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: how does this affect the indexing of rows then?

Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: ah, forget it -- that's in the next paragraph.

File: source/dofs/dof_tools.cc [hunk 5]
```diff
@@ -2067,8 +2067,15 @@ namespace DoFTools
...
-    for (cell=dof_handler.begin(level); cell != endc; ++i, ++cell)
+
+    for (cell=dof_handler.begin(level); cell != endc; ++cell)
+      if (cell->is_locally_owned())
+        ++i;
+    block_list.reinit(i, dof_handler.n_dofs(), dof_handler.get_fe().dofs_per_cell);
+    i=0;
+    for (cell=dof_handler.begin(level); cell != endc; ++cell)
       {
+        if (!cell->is_locally_owned()) continue;
```
Comment (bangerth) [inline source/dofs/dof_tools.cc]: I think it would be easier to read if you wrote it as
```
  for (cell=...)
    if (cell->is_locally_owned())
      {
         ....
```

File: tests/dofs/block_list_parallel_01.cc [hunk 6]
```diff
@@ -0,0 +1,62 @@
+// ---------------------------------------------------------------------
+//
```
Comment (bangerth) [inline tests/dofs/block_list_parallel_01.cc]: output files are missing for this test

File: include/deal.II/dofs/dof_tools.h (hunk 2 above)
Comment (guidokanschat) [inline include/deal.II/dofs/dof_tools.h]: Yes. The rows of the sparsity pattern relate to the locally owned cells. The size of the rows are always dofs per cell. Additionally, we have to reinit the sparsity pattern inside the function (like in make_vertex_patches), since the number of patches is determined inside the function.

File: include/deal.II/dofs/dof_tools.h (hunk 3 above)
Comment (guidokanschat) [inline include/deal.II/dofs/dof_tools.h]: It's not correct.

File: source/dofs/dof_tools.cc (hunk 5 above)
Comment (guidokanschat) [inline source/dofs/dof_tools.cc]: I don't think so, since the complexity rises

Comment (davydden) [inline source/dofs/dof_tools.cc]: i second Wolfgang's suggestion, IMHO it reads better.

**Comment by tamiko** [comment]
@guidokanschat Can you please add a small entry to ```doc/news/changes.h``` in the incompatible changes section?

File: include/deal.II/dofs/dof_tools.h (hunk 2 above)
Comment (tamiko) [inline include/deal.II/dofs/dof_tools.h]: My comment is just for context (to follow the discussion more easily):

We used to instantiate this function for (```template-arguments.in```):
```
SPARSITY_PATTERNS := { SparsityPattern;
                       DynamicSparsityPattern;
                       @DEAL_II_EXPAND_TRILINOS_SPARSITY_PATTERN@;

                       BlockSparsityPattern;
                       BlockDynamicSparsityPattern;
                       @DEAL_II_EXPAND_TRILINOS_BLOCK_SPARSITY_PATTERN@; }
```
Now, it is restricted to the plain ```SparsityPattern``` type.

Lacking any further insight into this particular function I'm curious why this change is necessary :-)

File: source/dofs/dof_tools.cc [hunk 7]
```diff
@@ -2067,8 +2067,15 @@ namespace DoFTools
     std::vector<types::global_dof_index> indices;
 
     unsigned int i=0;
-    for (cell=dof_handler.begin(level); cell != endc; ++i, ++cell)
+
+    for (cell=dof_handler.begin(level); cell != endc; ++cell)
+      if (cell->is_locally_owned())
+        ++i;
+    block_list.reinit(i, dof_handler.n_dofs(), dof_handler.get_fe().dofs_per_cell);
```
Comment (tamiko) [inline source/dofs/dof_tools.cc]: Is it because of this ```reinit```?

File: include/deal.II/dofs/dof_tools.h (hunk 2 above)
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: So is it correct that you want to use only `SparsityPattern` because this clearly is the most efficient format (you know the exact number of nonzeros) and nobody in their right mind would want to use anything else?

Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: Which, by the way, will currently only work for `::DoFHandler`, but not the hp variant. Of course, it's also not instantiated for `hp::DoFHandler` (and wouldn't compile), so maybe remove the `DoFHandlerType` template as well, while you're there? (I suspect that it was a template from back when we still had `MGDoFHandler`.

File: include/deal.II/dofs/dof_tools.h [hunk 8]
```diff
@@ -1699,18 +1699,27 @@ namespace DoFTools
...
-   * data structure is conveniently represented by a SparsityPattern object.
+   * Creates a sparsity pattern, which lists
+   * the degrees of freedom associated to each cell on the given
+   * level. This pattern can be used in RelaxationBlock classes as
+   * block list for additive and multiplicative Schwarz methods.
    *
-   * @note The ordering of rows (cells) follows the ordering of the standard
-   * cell iterators.
+   * The row index in this pattern is the cell index resulting
+   * from standard iteration through the Triangulation. For a
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: maybe say "through ONE LEVEL OF the triangulation"?

File: include/deal.II/dofs/dof_tools.h [hunk 9]
```diff
@@ -1699,18 +1699,27 @@ namespace DoFTools
...
+   * parallel::distributed::Triangulation, only locally owned cells
+   * are entered.
+   *
+   * The sparsity pattern is resized in this function to contain as
+   * many rows as there are locally owned cells on a given level, as
+   * many columns as there are degrees of freedom on this level.
+   *
+   * <tt>selected_dofs</tt> is a vector indexed by the local degrees
+   * of freedom on a cell. If it is used, only such dofs are entered
+   * into the block, which are selected. This allows for instance the
```
Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: block -> sparsity pattern
no comma

File: include/deal.II/dofs/dof_tools.h (hunk 2 above)
Comment (guidokanschat) [inline include/deal.II/dofs/dof_tools.h]: It can be made work for `hp::DoFHandler` by using the reinit with row length vector at a moderate performance and memory penalty

Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: Right, but the `selected_dofs` argument then no longer makes any sense.

Comment (guidokanschat) [inline include/deal.II/dofs/dof_tools.h]: Good point. One more reason for the change

File: tests/dofs/block_list_parallel_01.cc (hunk 6 above)
Comment (tamiko) [inline tests/dofs/block_list_parallel_01.cc]: @bangerth ```block_list_parallel_01.cc``` has a corresponding output file ```block_list_parallel_01.with_p4est=true.mpirun=2.output``` attached to this PR.

**Comment by tamiko** [comment]
:+1:

File: include/deal.II/dofs/dof_tools.h [hunk 10]
```diff
@@ -1699,19 +1699,28 @@ namespace DoFTools
...
+   * The sparsity pattern is resized in this function to contain as
+   * many rows as there are locally owned cells on a given level, as
+   * many columns as there are degrees of freedom on this level.
+   *
+   * <tt>selected_dofs</tt> is a vector indexed by the local degrees
+   * of freedom on a cell. If it is used, only such dofs are entered
+   * into the block list which are selected. This allows for instance
+   * the exclusion of components or of dofs on the boundary.
+   */
+  template <int dim, int spacedim=dim>
```
Comment (Rombur) [inline include/deal.II/dofs/dof_tools.h]: This requires C++11 http://cdash.kyomu.43-1.org/viewBuildError.php?onlydeltap&buildid=484

Comment (bangerth) [inline include/deal.II/dofs/dof_tools.h]: The default is also unnecessary because `spacedim` is deducible. @guidokanschat -- can you fix?
